File: kotlin/lesson 5/Trie.kt
// https://leetcode.com/problems/implement-trie-prefix-tree/

class Trie() {
    private val root = TrieNode()

    fun insert(word: String) {
        var node = root;
        for (c in word) {
            val index = c - 'a'
            if (node[index] == null) {
                node[index] = TrieNode()
            }
            node = node[index]!!
        }
        node.isWord = true
    }

    fun search(word: String) = findNode(word)?.isWord ?: false

    fun startsWith(prefix: String) = findNode(prefix) != null
    
    private fun findNode(prefix: String): TrieNode? {
        var node = root;
        for (c in prefix) {
            val index = c - 'a'
            if (node[index] == null) return null
            node = node[index]!!
        }
        return node
    }

    private class TrieNode {
        val children = Array<TrieNode?>(26, { null }) // a to z
        var isWord = false

        operator fun get(index: Int) = children[index]
        operator fun set(index: Int, value: TrieNode) { children[index] = value }
    }
}

File: javascript/lesson 5/trie.js
// https://leetcode.com/problems/implement-trie-prefix-tree/

class Trie {
    constructor() {
        this._root = new TrieNode();
    }

    insert(word) {
        let node = this._root;
        for (let c of word) {
            if (!node.get(c)) {
                node.set(c, new TrieNode());
            }
            node = node.get(c);
        }
        node.isWord = true;
    }

    search(word) {
        const node = this._findNode(word);
        return !!node && node.isWord;
    }

    startsWith(prefix) {
        const node = this._findNode(prefix);
        return !!node;
    }

    _findNode(prefix) {
        let node = this._root;
        for (let c of prefix) {
            node = node.get(c);
            if (!node) return null;
        }
        return node;
    }
}

class TrieNode {
    constructor() {
        this.children = new Array(26); // a to z
        this.isWord = false;
    }

    get(c) {
        return this.children[c.charCodeAt(0) - 'a'.charCodeAt(0)];
    }

    set(c, value) {
        this.children[c.charCodeAt(0) - 'a'.charCodeAt(0)] = value;
    }
}
